File: GomokuAIJavaProject/AIGomoku/GameManager.java
package AIGomoku;

import java.util.Scanner;

public class GameManager {
	
	
	public static void launchGame() {
		try (Scanner scan = new Scanner(System.in)) {
			System.out.println("Select a game mod");
			System.out.println("1. Two player");
			System.out.println("2. One player vs IA");
			System.out.println("3. Two IA");
			int code = scan.nextInt(); 
			while(code != 1 && code != 2 && code != 3) {
				System.out.println("Select a valid game mod");
				System.out.println("1. Two player");
				System.out.println("2. One player vs IA");
				System.out.println("3. Two IA");
				code = scan.nextInt();
			}
			if(code == 1) {
				//launch2playerMode
				BoardGame board = BoardGame.getInstance();
				Controller controller = Controller.getInstance();
				boolean gameFinish = false;
				boolean movePlayed = false;
				int turn = 1;
				int posX,posY;

				//boucle de jeu
				while(!gameFinish) {
					System.out.println("Player " + turn + " turn");
					while(!movePlayed) {
						System.out.println("Enter position x where you want to place the token");
						posX = controller.getPos();
						System.out.println("Enter position y where you want to place the token");
						posY = controller.getPos();
						movePlayed = board.placeToken(turn, posX, posY); 
						if(board.isFinished(new Position(posY,posX),turn)) {
							gameFinish = true;
						}
						board.getBoard().printBoard();
						if(!movePlayed) {
							System.out.println("Move not valid");
						}
					}
					if(gameFinish) {
						System.out.println("Player " + turn + " Win !");
					}
					
					movePlayed = false;
				}
			}
			else if(code == 2) {
				//Launch One player vs IA mode
				BoardGame board = BoardGame.getInstance();
				Controller controller = Controller.getInstance();
				int value[] = {1,10,100,1000,10000,100000};
				AIGomoku AI = new AIOpen((char) 2,4,new ElaborateEvaluateFunction(value));
				boolean playerTurn = false;
				boolean gameFinish = false;
				boolean movePlayed = false;
				int posX;
				int posY;
				while(!gameFinish) {
					if(playerTurn) {
						System.out.println("Player turn");
						while(!movePlayed) {
							System.out.println("Enter position x where you want to place the token");
							posX = controller.getPos();
							System.out.println("Enter position y where you want to place the token");
							posY = controller.getPos();
							movePlayed = board.placeToken(1, posX, posY); 
							if(board.isFinished(new Position(posY,posX),1)) {
								gameFinish = true;
							}
							if(!movePlayed) {
								System.out.println("Move not valid");
							}
							movePlayed = true;
						}
					}
					else {
						System.out.println("IA turn");
						gameFinish = AI.playMove(board);

					}
					board.getBoard().printBoard();

					playerTurn = !playerTurn;
					
					movePlayed = false;
					
				}
				if(playerTurn) {
					System.out.println("Player Win");
				}
				else{
					System.out.println("AI Win");
				}
				playerTurn = !playerTurn;
			}
			else {
				// Launch Two IA mode
				BoardGame board = BoardGame.getInstance();
				int value[] = {1,2,5,10,20,5000,100000};
				AIGomoku AI = new AIOpen((char)1,4,new JumpEvaluateFunction(value));
				AIGomoku AItwo = new RandomAIGomoku(2);
				int AITurn = 2;
				System.out.println(AITurn);
				
				System.out.println("AITurn is : " + AITurn);
				boolean gameFinish = false;
				while(!gameFinish) {
					if(AITurn == 1) {
						gameFinish = AItwo.playMove(board);
						AITurn++;
					}
					else {
						gameFinish = AI.playMove(board);
						AITurn--;
					}
					board.getBoard().printBoard();
				}
				if(AITurn == 2) {
					System.out.println("AI two win");
				}
				else {
					System.out.println("AI one win");
				}
			}
		}
		System.out.println("Number of move played is : " + BoardGame.numberOfMovesPlayed);
       }
     
	
}

File: CPP/src/GameManager.cpp
#include "GameManager.hpp"

void GameManager::LaunchGame()
{
    printMenu();
    unsigned int choice;
    std::cin >> choice;
    while(choice != PLAYER_VS_PLAYER && choice != PLAYER_VS_IA && choice != IA_VS_IA){
        std::cout << "Choose a valid Number" << std::endl;
        printMenu();
        std::cin >> choice;
    }

    switch(choice)
    {
        case PLAYER_VS_PLAYER:
        {
            this->playerVersusPlayer();
            break;
        }
        case PLAYER_VS_IA:
        {
            this->playerVersusIA();
            break;
        }
        case IA_VS_IA:
        {
            this->IAVersusIA();
            break;
        }
    }
}

void GameManager::printMenu()
{
    std::cout << "Select a game mod"   << "\n" 
              << "1. Two player"       << "\n" 
			  << "2. One player vs IA" << "\n"
			  << "3. Two IA"           << std::endl;
}

void GameManager::playerVersusPlayer()
{
    BoardGame* board = BoardGame::getInstance();
    bool gameFinish = false;
    bool movePlayed = false;
    idPlayer turn = idPlayer::p1;
    Position* pos = new Position(0,0);
    board->getBoard()->printBoard();
    //boucle de jeu
    while(!gameFinish) {
        std::cout << "Player " << turn << " turn" << std::endl;
        while(!movePlayed) {
            std::cout << "Enter position x where you want to place the token" << std::endl;
            pos->setX(Controller::getPos());
            std::cout << "Enter position y where you want to place the token" << std::endl;
            pos->setY(Controller::getPos());
            movePlayed = board->placeToken(turn,pos); 
            board->getBoard()->printBoard();
            std::cout << "x is : " << pos->getX() << " y is : " << pos->getY() << "and move is : " << movePlayed << std::endl;

            if(!movePlayed) {
                std::cout << "Move not valid" << std::endl;
            }
            else{
                if(board->isFinished(pos,turn)) {
                    gameFinish = true;
                }
            }
        }
        if(gameFinish) {
            std::cout<< "Player " << turn << " Win !" << std::endl;
        }
        if(turn == idPlayer::p2){
            turn = idPlayer::p1;
        }
        else{
            turn = idPlayer::p2;
        }
        movePlayed = false;
    }
}

void GameManager::playerVersusIA()
{
    //Launch One player vs IA mode
    BoardGame* board = BoardGame::getInstance();
    int value[] = {1,10,100,1000,10000,100000,1000000};
    int value2[] = {1,10,100,1000,10000,100000};
    BasicEvaluateFunction* func = new BasicEvaluateFunction();
    ElaborateEvaluateFunction* func2 = new ElaborateEvaluateFunction(value);
    JumpEvaluateFunction* func3 = new JumpEvaluateFunction(value2);
    AIGomoku* AI = new AIOpen(idPlayer::p2,5,func2);
    idPlayer playerTurn = idPlayer::p1;
    bool gameFinish = false;
    bool movePlayed = false;
    int posX;
    int posY;
    Position* pos = new Position(0,0);
    board->getBoard()->printBoard();
    while(!gameFinish) {
        if(playerTurn == idPlayer::p1) {
            std::cout << "Player turn" << std::endl;
            while(!movePlayed) {
                std::cout << "Enter position x where you want to place the token" << std::endl;
                pos->setX(Controller::getPos());
                std::cout << "Enter position y where you want to place the token" << std::endl;
                pos->setY(Controller::getPos());
                movePlayed = board->placeToken(playerTurn,pos); 
                board->getBoard()->printBoard();
                std::cout << "x is : " << pos->getX() << " y is : " << pos->getY() << "and move is : " << movePlayed << std::endl;

                if(!movePlayed) {
                    std::cout << "Move not valid" << std::endl;
                }
                else{
                    if(board->isFinished(pos,playerTurn)) {
                        gameFinish = true;
                    }
                }
            }
        }
        else {
            std::cout << "IA turn" << std::endl;
            gameFinish = AI->playMove(board);

        }
        board->getBoard()->printBoard();

        if(playerTurn == idPlayer::p2){
            playerTurn = idPlayer::p1;
        }
        else{
            playerTurn = idPlayer::p2;
        }
        movePlayed = false;
        
    }
    if(playerTurn) {
        std::cout << "Player Win" << std::endl;
    }
    else{
        std::cout << "AI Win" << std::endl;
    }
}

void GameManager::IAVersusIA()
{

}